### PR TITLE
Tests and semaphoreci integration

### PR DESCRIPTION
**Pull request by soerenmartius**
Implement an integration test:

- Create ECR repository
- Build a docker image
- Authenticate with ECR
- Push the docker image to ECR
-  Use Docker-out-of-Docker ( DooD ) approach to build docker image inside the build-tools container

### CODE REVIEW

**Comment by soerenmartius** [comment]
Hi Marius,

The code is still a bit messy but it should be clear what we're doing here: 
- Create ECR repository
- Build a docker image
- Authenticate with ECR
- Push the docker image to EC

Executing the test with `AWS_ACCESS_KEY_ID=XXXX AWS_SECRET_ACCESS_KEY=XXXX go test -v -parallel 128  ./...     ` works well now.

However, we currently can't run the test with our (https://hub.docker.com/r/mineiros/build-tools)[build-tools} docker image. If we'd like to be able to run the test inside a docker container, we need to add docker as a dependency to the build-tools image and start a privileged container. I don't think that this is a good practice though.

I would suggest, that we use a packer image to create and push the test docker image. However, that would come with a few downsides:
- We can't cache docker builds with packer
- The test would take longer
- The test would likely create costs because packer spans up an EC2 instance.

wdyt @mariux ?

**Comment by soerenmartius** [comment]
> lgtm

We shouldn't approve PR's with failing tests though...

**Comment by soerenmartius** [comment]
> Hi Marius,
> 
> The code is still a bit messy but it should be clear what we're doing here:
> 
> * Create ECR repository
> * Build a docker image
> * Authenticate with ECR
> * Push the docker image to EC
> 
> Executing the test with `AWS_ACCESS_KEY_ID=XXXX AWS_SECRET_ACCESS_KEY=XXXX go test -v -parallel 128 ./... ` works well now.
> 
> However, we currently can't run the test with our ([https://hub.docker.com/r/mineiros/build-tools)[build-tools}](https://hub.docker.com/r/mineiros/build-tools)%5Bbuild-tools%7D) docker image. If we'd like to be able to run the test inside a docker container, we need to add docker as a dependency to the build-tools image and start a privileged container. I don't think that this is a good practice though.
> 
> I would suggest, that we use a packer image to create and push the test docker image. However, that would come with a few downsides:
> 
> * We can't cache docker builds with packer
> * The test would take longer
> * The test would likely create costs because packer spans up an EC2 instance.
> 
> wdyt @mariux ?

I, for now, took the easiest way possible: [Docker out of  Docker](http://tdongsi.github.io/blog/2017/04/23/docker-out-of-docker/)

**Comment by soerenmartius** [comment]
Ready for review @mariux

**Comment by soerenmartius** [comment]
I just saw that whenever we use `terraform output`, terratest writes the full output to stdout. It's not critical, since the created IAM User comes without any implicit permissions and will be destroyed after the test finished. 
Still, it's not acceptable and a no go. I am gonna open a PR for terratest to suppress output logs for sensitive outputs.

**Comment by soerenmartius** [comment]
> I just saw that whenever we use `terraform output`, terratest writes the full output to stdout. It's not critical, since the created IAM User comes without any implicit permissions and will be destroyed after the test finished.
> Still, it's not acceptable and a no go. I am gonna open a PR for terratest to suppress output logs for sensitive outputs.

https://github.com/gruntwork-io/terratest/issues/476

File: outputs.tf [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+output "repository_id" {
+  description = "The Id of the repository."
+  value       = try(aws_ecr_repository.repository[0].id, null)
+}
+
+output "repository_url" {
+  description = "The URL of the repository."
+  value       = try(aws_ecr_repository.repository[0].repository_url, null)
+}
+
```
Comment (mariux) [inline outputs.tf]: why? you can get them from the repository itself
```
module.name.repository.arn
module.name.repository_arn
```

Comment (soerenmartius) [reply outputs.tf]: fixed in aaa3cc7

Comment (soerenmartius) [reply outputs.tf]: My take on outputs:

I don't have a strong opinion yet but here are some bullet points that lead me in the direction of explicit outputs:

- returning the whole object doesn't really create a contract / valid API. What if the internal representation changes due to Provider upgrade?  The module won't be able to guarantee backward compatibility anymore. If we return explicit outputs we can guarantee BC.
- What does the user really want? Returning the whole resource will most likely end in a scenario in which the user only uses 20% of the outputs anyways ... do we really need all information?
- We might return sensitive outputs per default if we always return the whole object?
- No convenient API - the user would need to check the terraform providers documentation to get an overview of the outputs.

I will post this in the related [trello task](https://trello.com/c/P4D2Y0BC/4-arrange-a-unified-output-format-for-all-terraform-modules) also. 
wdyt?

Also: I think we might be ready to get this merged now ? :)

**Pull request merged**